eng/testing/debug-dump-template.md

```diff
@@ -0,0 +1,126 @@
+# Debugging a CI dump
+
+This document describes how to debug a CI/PR test dump by downloading assets from helix, using a dotnet tool called `runfo`.
+
+## What is runfo?
+
+Runfo is a dotnet global tool that helps get information about helix test runs and azure devops builds. For more information see [this](https://github.com/jaredpar/runfo/tree/master/runfo#runfo)
+
+### How do I install it?
+
+You just need to run:
+
+```script
+dotnet tool install --global runfo
+```
+
+If you already have it installed, make sure you have at least version `0.6.1` installed, which contains support to download helix payloads. If you don't have the latest version just run:
+
+```script
+dotnet tool update --global runfo
+```
+
+## Download helix payload containing symbols:
+
+You can just achieve this by running:
+
+```script
+runfo get-helix-payload -j %JOBID% -w %WORKITEM% -o <out-dir>
```

Comment:
Can you outline from where someone gets the relevant jobid and workitem?

Reply:
This will be filled in by the python script. So the one that will show up by the dump will contain that info there already, that's why I didn't put that info. I was going to put it but then when I so the produced file it didn't make sense. Thoughts on that?

Reply:
You're saying I literally write `-j %JOBID% -w %WORKITEM%` rather than supplying those values myself?  Why do I even need to write that then, i.e. why is the command not just `runfo get-helix-payload -o <out-dir>`?

Reply:
No not really. What I'm saying is that this is just a template that we feed into the python script that I'm adding as part of this PR and the python script will fill in this information and produce a .md file with all the information in it. For example, this .md file was generated from a crash I forced: https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-43937-merge-229762be8cb749d196/Microsoft.CSharp.Tests/debug-dump.md?sv=2019-07-07&se=2020-11-18T23%3A31%3A30Z&sr=c&sp=rl&sig=YzLMjEzJP8W3zhGSp5m2JVzphr%2FgNVm0xRO%2FnV4wNJk%3D

You can see that the command it has is: 
```
runfo get-helix-payload -j 229762be-8cb7-49d1-960a-83e041494e15 -w a760fa7c-e088-41ca-b565-640db87df271 -o <out-dir>
```

So what I was trying to say, is that if you just read the template checked in into the repo, it makes sense to add the information from where to get the jobid... however the purpose of this template is to produce a .md file that is side by side with the dump in the helix logs container, so in the final .md file it will be confusing to have a section to explain "how to get the helix jobid" when you don't even need it since the jobid is already embedded into the .md file you're reading. 

Does that makes it more clear?

Reply:
Sorry, I'm still not understanding.  The directions say all I have to do is:
```
dotnet tool install --global runfo
runfo get-helix-payload -j %JOBID% -w %WORKITEM% -o outputdir
```
How does it know what job I care about / what to download?  So is there a step missing from the instructions?

Reply:
This is a template. It gets populated before it the upload to the AzDO tab.

Reply:
Populated with what?  I haven't told it what I care about.

I'm confused.  I want to debug an issue that occurred in CI (which is what this doc is helping me with).  What exactly do I type at the command line?  Let's say my PR for dotnet/runtime PR 12345 failed.  What exactly do I do?

Reply:
It will be populated with the actual helix job id and workitem id for the crash that you hit in CI.

We populate it here: https://github.com/dotnet/runtime/pull/43937/files#diff-fd8bb0da491f288f6e268f835a49b14126ba3f1a15ab6971ddc76dd480b81b11R95

We pass down that information here: https://github.com/dotnet/runtime/pull/43937/files#diff-1d68a522b60c1ab69fb537ce06f5d7e9a176fa4f75e16425fc164d4b738d7678R216

> What exactly do I type at the command line? Let's say my PR for dotnet/runtime PR 12345 failed. What exactly do I do?

You will go to the azure devops test tab which will show the failure and then in the helix attachments you will see the dump and a MarkDown file. That MD file, will be this template but with the info that you need embedded in there so that you don't have to go and get the helix job yourself. It will look like this:

![image](https://user-images.githubusercontent.com/22899328/97765068-f79fe080-1acd-11eb-9cae-82f91418f382.png)

So the steps you'll follow is download the dump and debug-dump.md file (which I will rename to how-to-debug-dump.md) and just follow the instructions in the MD file, no need to get more information. This is trying to minimize the steps people have to do and info they need to gather whenever they see a dump in their test results. The final MD file will look like [this one](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-43937-merge-229762be8cb749d196/Microsoft.CSharp.Tests/debug-dump.md?sv=2019-07-07&se=2020-11-18T23%3A31%3A30Z&sr=c&sp=rl&sig=YzLMjEzJP8W3zhGSp5m2JVzphr%2FgNVm0xRO%2FnV4wNJk%3D). It will even add a link to the build artifacts in case you need to download the cross DAC to debug a linux dump on Windows. 

> Go to your build by navigating to [your build](https://dnceng.visualstudio.com/public/_build/results?buildId=870703&view=artifacts&type=publishedArtifacts) and download the `CoreCLRCrossDacArtifacts`, then choose the cross DAC you need depending on the flavor.

Reply:
I see.  I totally misunderstood the purpose of this documentation. It's not documentation for someone to read in the dotnet/runtime repo.  It's effectively part of our CI tooling. Thanks :)

Reply:
> I see. I totally misunderstood the purpose of this documentation. It's not documentation for someone to read in the dotnet/runtime repo. It's effectively part of our CI tooling. Thanks :)

Yeah, that's why I put it under `eng\testing`... however, it can serve as documentation as well, but it will just miss getting this piece of information. I could add it and then strip it out when generating the template if you'd like that?

Reply:
> I could add it and then strip it out when generating the template if you'd like that?

Nah, it's fine, but thanks.